### PR TITLE
iojs v1.0.3 (new formula)

### PR DESCRIPTION
**Pull request by DomT4**
Alternative implementation of #35853 for discussion/giggles/etc.

### CODE REVIEW

File: Library/Formula/iojs.rb [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    temporarily shunt iojs's Node symlink to the front of your $PATH.
+    This allows npm to find the desired version of Node rather than
+    any other version of Node available along the $PATH, and consequently
+    you can directly install modules with `npm-iojs install -g grunt`.
+
+    However, your modules installed via `npm` do not currently reside
+    in your `$PATH` as there's a limited amount Homebrew can do to isolate
+    those. If you wish to use those modules you should permanently
+    prepend your `$PATH` like such:
+      export PATH=#{HOMEBREW_PREFIX}/share/iojs/bin:#{libexec}/iojs/bin:$PATH
```
Comment (DomT4) [inline Library/Formula/iojs.rb]: On this line: If you follow that line, you can just do ` npm install -g blah ` or similar. We could simplify this, but I wanted to give maintainers and people options for discussion for now, and I wanted to give developers a way to activate the iojs npm without having it absolutely conflict with the ` Node ` npm. This reasonably-hideous workaround achieves that.

**Comment by DomT4** [comment]
Further discussion: Yes, for now, I've removed the option to *not* install ` npm `. Honestly, not installing `npm` in this particular format & formula is a complete pain in the neck, and the scripts become much less useful without ` npm ` around. I'll put it back if there's strong demand for the option from the maintainers or well, anyone really, but at least for testing purposes I think mandating it isn't the most awful thing we could do.

**Comment by tdsmith** [comment]
Can you summarize the major points of difference with the existing PR?

**Comment by DomT4** [comment]
Sure.

1) The `npm` option has been cut. It's mandatory. Reasoning is in the post above.

2) This formula installs iojs and npm into ` libexec ` and then moves `npm` into ` homebrew_prefix/share/iojs/lib/node_modules ` and creates exec scripts for `npm` and `iojs` into the main `homebrew_prefix/bin`. This allows it to not conflict with Node whilst being more readily usable and accessible than ` keg_only ` or a solid ` conflicts_with ` does.

The exec scripts are the kinda nifty part to play with here - the `iojs ` exec script triggers an instruction to `iojs` so it knows where the `node_modules` for it are located, and the `npm` exec script triggers a temporary jiggle to the `$PATH` so `npm` can always find the `node` symlink from iojs.

3) The new `iojs/npm_modules` bin becomes where `npm` executables are placed, and the overly lengthy caveat that we'd need to trim over time explains how to access those. What we could do, over a longer-term, is write two short bash scripts that would allow users to do `gonode` or `goiojs` and it would switch the `node` symlink and `node_path` over automatically. This wouldn't be difficult, but it's not without risk, so I considered a smaller step a better way to get there.

4) As alluded to above, ` npm install -g example ` installs `example` into ` HOMEBREW_PREFIX/share/iojs/lib/node_modules ` and symlinks that to ` HOMEBREW_PREFIX/share/iojs/bin `.

**Comment by DomT4** [comment]
On another small note: I know @othiym23 wanted to avoid renaming `npm` away to something else, and this is a compromise of that. The exec script placed in ` homebrew_prefix/bin ` is called ` npm-iojs ` but it is really just a glorified symlink for all intents and purposes - It's also something you can easily negate having to use by prepending the ` $PATH ` as discussed in the ` caveats `. If you prepend the ` $PATH ` you can just do ` npm install -g example ` and that would work.

File: Library/Formula/iojs.rb [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+    cp_r libexec/"npm/lib/node_modules/npm", iojs_modules
+    ln_sf npm_exec, iojs_bindir/"npm"
+
+    # Let's do the manpage dance. It's just a jump to the left.
+    # And then a step to the right, with your hand on rm_f.
+    ["man1", "man3", "man5", "man7"].each do |man|
+      mkdir_p HOMEBREW_PREFIX/"share/man/#{man}"
+      rm_f Dir[HOMEBREW_PREFIX/"share/man/#{man}/{npm.,npm-,npmrc.}*"]
+      Dir[libexec/"npm/share/man/#{man}/npm*"].each { |f| ln_sf f, HOMEBREW_PREFIX/"share/man/#{man}" }
```
Comment (xu-cheng) [inline Library/Formula/iojs.rb]: https://github.com/Homebrew/homebrew/commit/cb5ace6797df483621020149eb495d1f61da1cd9

Comment (DomT4) [inline Library/Formula/iojs.rb]: Nice spot! Will fix as soon as my internet decides it wants to go a little faster than 7KB/s :crying_cat_face:

File: Library/Formula/iojs.rb [hunk 3]
```diff
@@ -0,0 +1,125 @@
+class Iojs < Formula
+  homepage "https://iojs.org/"
+  url "https://iojs.org/dist/v1.0.3/iojs-v1.0.3.tar.gz"
+  sha256 "c7fe7f71d9920f4cfd930f9022c7dddca7e11a0377fed1ee23e3241c2952db42"
+
+  option "with-debug", "Build with debugger hooks"
+  option "with-completion", "install bash completion for npm"
```
Comment (mikemcquaid) [inline Library/Formula/iojs.rb]: Let's make this `without-completion`

Comment (DomT4) [inline Library/Formula/iojs.rb]: I didn't know if that was too clashing against Node, given this deliberately removes the conflict and keg mechanism? *(As an experiment. Not saying this is the best way to get it done, just an alternative to the other discussion)*. I suppose I could do a little if file already exists don't replace it detection dance?

**Comment by DomT4** [comment]
Incredibly useful upstream list monitoring what's broken in terms of packages and when they are fixed [here](https://github.com/iojs/io.js/issues/456). Thanks to @timoxley for setting that up.

**Comment by DomT4** [comment]
Per Mike's comment in the other thread, closing this PR :+1:
